[IMP] account_chart_update: let user choose digits

yajo opened the pull request:
Before this change, if somebody decided to use a different amount of digits for their account codes, every time the wizard got executed it detected every account as something to change.

Also, if someone already had created a valid account with a different amount of digits, when executing the wizard it wouldn't detect it as a valid match because of this reason.

FTR, digits are hardcoded in the chart template. This is usually done by Odoo and sometimes the digits chosen is somewhat arbitrary.

After this change, the user can now change the desired amount of digits generated by the wizard.

![imagen](https://user-images.githubusercontent.com/973709/209550431-52a2b088-bdd8-48ed-b9e1-83191e30cadf.png)


Also, there's a new option for matching accounts. It is called "code prefix". It just detects that the account has the same code as the account template, followed by any amount of zeros. This new option is enabled by default, with lower priority than the "code" option.

![imagen](https://user-images.githubusercontent.com/973709/209550487-36b18bdb-ac3b-4137-99a1-1e2de09bcdbb.png)


OTOH, only managers will be able to execute the chart of account wizard. This reflects [the requirement hardcoded upstream][1].

Summarizing, now you can choose to have a different amount of digits in your accounts, and that'll be just OK.

@moduon MT-1923

[1]: https://github.com/odoo/odoo/blob/4ac5ec6a2c4b3e2707446acb1cd693808f0d3c8a/addons/account/models/chart_template.py#L234

pedrobaeza commented:
This was consciously removed from both Odoo core and the module `account_chart_update` to avoid people doing bad practices changing the number of digits, as there's no need of more than 6 digits. If you have the need in a migrated database, just go to the `account.chart.template` record, and modify the fields indicating the number of digits, or put this as an extra module on top of `account_chart_update`, but please don't contaminate the current module with something that has been ripped-off on purpose. You can add the hooks you need on the base module though.

rafaelbn commented:
Please, could you point us to that decisions? Just for sharing. I have been looking for it but I can find only [ [FIX] account_chart_update: Fetch account code digits correctly #566 ](https://github.com/OCA/account-financial-tools/pull/566)

I don't want obviously discuss (or lost time) about who say that something in accounting is bad practice, accountants, auditors or developers. Or the need or not the need. Depends in many factors.

The objetive of the task is have the ability of update account chart update in any contition. It doesn't mind how many digit the accountants of the client have, it minds about updating chart without problems and as fast as possible.

We can solve it outside OCA obviusly but maybe is a good help and tool.

@legalsylvain @fclementic2c can you share how many digits accountants have in the Odoo database you maintain? Just curiosity 😄 

HAPPY NEW YEAR 2023! 🥳

yajo commented:
The possibility of choosing the number of digits was removed for v12 upstream in https://github.com/odoo/odoo/pull/16587.

yajo commented:
@pedrobaeza this PR includes 3 enhancements:

1. Require some permissions to open the wizard.
3. Improve matching by supporting code prefix. This way, if you import a DB that includes an account with code `4750000000000`, it will be a correct match for the account `475000` that the wizard generates.
4. Let user choose number of digits.

Do you consider all 3 enhancements "contamination"?

pedrobaeza commented:
The number 3 is the one that should be extracted in a separate module.

yajo committed:
[IMP] account_chart_update: let user know digits

Before this change, if somebody decided to use a different amount of digits for their account codes, every time the wizard got executed it detected every account as something to change.

Also, if someone already had created a valid account with a different amount of digits, when executing the wizard it wouldn't detect it as a valid match because of this reason.

FTR, digits are hardcoded in the chart template. This is usually done by Odoo and sometimes the digits chosen is somewhat arbitrary.

After this change, the user can now know the desired amount of digits before generating them by the wizard. Also, there's a new option for matching accounts. It is called "code prefix". It just detects that the account has the same code as the account template, followed by any amount of zeros. This new option is enabled by default, with lower priority than the "code" option.

OTOH, only managers will be able to execute the chart of account wizard. This reflects [the requirement hardcoded upstream][1].

Summarizing, now you can know the amount of digits in your accounts, and that'll be just OK.

@moduon MT-1923

[1]: https://github.com/odoo/odoo/blob/4ac5ec6a2c4b3e2707446acb1cd693808f0d3c8a/addons/account/models/chart_template.py#L234

yajo commented:
OK, the PR is now the same, but users cannot edit the code digits from within the wizard.

pedrobaeza commented:
But if the logic is still there, it's a maintenance burden. Please extract all the digits thing to a separate module.

yajo commented:
I removed the point 3 from above, as you said. What do you mean by "the logic is still there"?

pedrobaeza commented on account_chart_update/security/ir.model.access.csv:
```diff
@@ -1,9 +1,9 @@
-access_wizard_update_charts_accounts_tax,wizard.update.charts.accounts.tax,model_wizard_update_charts_accounts_tax,,1,1,1,1
-access_wizard_matching,wizard.matching,model_wizard_matching,,1,1,1,1
-access_wizard_update_charts_accounts_fiscal_position,wizard.update.charts.accounts.fiscal.position,model_wizard_update_charts_accounts_fiscal_position,,1,1,1,1
+access_wizard_update_charts_accounts,wizard.update.charts.accounts,model_wizard_update_charts_accounts,base.group_erp_manager,1,1,1,1
```
This should be account_manager, not ERP manager (for all).

yajo replied:
When you load the COA this permission requirement is hardcoded in https://github.com/odoo/odoo/blob/4ac5ec6a2c4b3e2707446acb1cd693808f0d3c8a/addons/account/models/chart_template.py#L234.

pedrobaeza reviewed (changes requested):
The pre-computed code prefix is not a good strategy, as depending on the accounts, on templates the prefix already contains 0s, and this is only going to make the module installation slower. Please use another technique. Why the on-the-fly filling is not enough?

yajo commented:
For the case when the template contains zeros, they are removed before using them to match the prefix. Please double-check the code. This is exercised implicitly in one of the tests.

Module installation times are often not a valid rejection argument, unless we have `O^n` complexity. Here we're just passing a very simple computation with `O*n` complexity, where the amount of records is not expected to amount for millions (just `account.account` records... maybe a few hundred in the worst case? A couple thousands? Not very problematic...).

For valid on-the-fly matching, I'd need to use a regexp similar to `f"^{prefix}0*$"`. Since we're matching database records, that means to use a PostgreSQL-level regexp. [Upstream docs](https://www.postgresql.org/docs/current/functions-matching.html) say:

> While most regular-expression searches can be executed very quickly, regular expressions can be contrived that take arbitrary amounts of time and memory to process. Be wary of accepting regular-expression search patterns from hostile sources. If you must do so, it is advisable to impose a statement timeout.
>
> Searches using SIMILAR TO patterns have the same security hazards, since SIMILAR TO provides many of the same capabilities as POSIX-style regular expressions.

We should consider that the code field in the `account.account.template` record is a hostile source, as it can be easily abused. I could circumvent that hazard by simply pre-validating it, but then it would mean to add more low-level stuff to this module (replacing domain criteria with direct SQL queries, or adding a search function that uses the hidden `inselect` operator), which is already quite complex.

Having all this into account, I consider that a hazard-free high-level easy-to-understand ORM-only solution is better, at the expense of having one extra harmless hidden computed field in the `account.account` model.

In the improbable case that there's a CPU leak in installing this module due to the compute we're adding, another possibility is to fill the field through SQL. But I'd only do that after there's enough proof that there's really a leak, to avoid unnecessary early optimization.

So AFAICS none of your rejection points is valid. Can we merge then?

BTW, if you feel like this feature is gonna be too much maintenance burden for you, I'll willingly volunteer to co-maintain this module. Less work for you, more features for me, we all win! 😃 

Thanks.

yajo committed:
[FIX] account_chart_update: responsive views

Make its 2nd step usable.

@moduon MT-1923

yajo commented:
I added a small new improvement. Makes 2nd step page go from this:

https://user-images.githubusercontent.com/973709/210229560-1abd2efa-4f7d-44fd-9d9f-bdf2e76afefc.mp4

to this:


https://user-images.githubusercontent.com/973709/210229578-12ebe37b-a0e7-43d2-923c-abd8ddaa18c7.mp4

FTR same problem as upstream https://github.com/odoo/odoo/pull/90946.

fcvalgar reviewed (approved):
LGTM, thank you @yajo.

rafaelbn reviewed (approved):
Tested in runboat I see this as an improvement 👍
